### DISCUSSION TITLE
Not able to install java in gocd agent

### DISCUSSION THREAD

**Discussion by Tanmoy037** [gocd/gocd#13204] [Q&A]
##### Summary

gocd/gocd-agent-wolfi:v24.3.0
- this agent image does not have mvn or java installed in it. I need a my gocd agent to build java springboot project. For which I need the agent to have java and mvn install in it.

I am trying to use this docker file to create a image for myself - 

```Dockerfile
FROM gocd/gocd-agent-wolfi:v24.3.0

USER root

# Install Java 21 JDK
RUN apk add --no-cache openjdk21-jdk

# Install Java 8
RUN apk add --no-cache openjdk8

# Set JAVA_HOME for Java 21
ENV JAVA_HOME=/usr/lib/jvm/java-21-openjdk
ENV PATH=$JAVA_HOME/bin:$PATH

# Add environment variable for Java 8 path (useful for switching)
ENV JAVA8_HOME=/usr/lib/jvm/java-8-openjdk

# Install other dependencies
RUN apk add --no-cache maven
RUN apk add --no-cache terraform
RUN apk add --no-cache python3 py-pip
RUN pip install awscli

# Create .aws directory if it doesn't exist
RUN mkdir -p /root/.aws

# Add AWS credentials
ADD ./aws.credentials /root/.aws/credentials
```

but giving this error - 
```
 > [2/9] RUN apk add --no-cache openjdk21-jdk-21.0.5_p11-r0:                                                                                                                                   
0.185 fetch https://packages.wolfi.dev/os/x86_64/APKINDEX.tar.gz                                                                                                                               
2.963 ERROR: unable to select packages:                                                                                                                                                        
2.963   openjdk21-jdk-21.0.5_p11-r0 (no such package):
2.963     required by: world[openjdk21-jdk-21.0.5_p11-r0]
------
Dockerfile:6
```

Please tell me if there is any mvn installed docker image for agent is available ? 
or help me to build a correct docker file. 
p.s -> using this image for server gocd/gocd-server:v24:3:0

**Comment by chadlwilson** [comment]
> Please tell me if there is any mvn installed docker image for agent is available ? or help me to build a correct docker file. p.s -> using this image for server gocd/gocd-server:v24:3:0

No, you have to create your own child-image or use another mechanism to provide whatever tools your build jobs/tasks need.

More modern practice for Maven is often to use the `mvnw` wrapper scripts that will download the right Maven version at runtime and avoid having to have it pre-installed.

> but giving this error -
> 
> ```
>  > [2/9] RUN apk add --no-cache openjdk21-jdk-21.0.5_p11-r0:                                                                                                                                   
> 0.185 fetch https://packages.wolfi.dev/os/x86_64/APKINDEX.tar.gz                                                                                                                               
> 2.963 ERROR: unable to select packages:                                                                                                                                                        
> 2.963   openjdk21-jdk-21.0.5_p11-r0 (no such package):
> 2.963     required by: world[openjdk21-jdk-21.0.5_p11-r0]
> ------
> Dockerfile:6
> ```

This is a Wolfi problem, not really related to GoCD. Not sure where you got `openjdk21-jdk` as the package name on Wolfi from, try `openjdk-21` (and `openjdk-8`)/

```
RUN apk add --no-cache openjdk-21
```

**Comment by Tanmoy037** [comment]
Thanks it solved the java, mvn issue. Can you recommend me for python and pip. Getting this error -

=> ERROR [5/7] RUN apk add --no-cache python3 py-pip                                                                                                                                     2.9s 
------                                                                                                                                                                                         
 > [5/7] RUN apk add --no-cache python3 py-pip:                                                                                                                                                
0.239 fetch https://packages.wolfi.dev/os/x86_64/APKINDEX.tar.gz                                                                                                                               
2.714 ERROR: unable to select packages:                                                                                                                                                        
2.714   py-pip (no such package):
2.714     required by: world[py-pip]
------

**Reply by chadlwilson** [reply]
`py3-pip`.

You need to find the appropriate package names/aliases for the wolfi distribution if you would like to use a wolfi base image. There are many ways to do package search, e.g use https://wolfi.pkgs.org/rolling/wolfi-base-x86_64/ and filter the distribution for Wolfi only, look at https://github.com/wolfi-dev/os

Or just use `apk search` directly:

```
docker run -it cgr.dev/chainguard/wolfi-base sh -c "apk update && apk search pip"
```

**Comment by Tanmoy037** [comment]
FROM gocd/gocd-agent-wolfi:v24.3.0

USER root

# Install Maven
RUN apk add --no-cache maven

# Install OpenJDK 8 and Java 21
RUN apk add --no-cache openjdk-8 openjdk-21

# Set JAVA_HOME to the default Java version (Java 21 in this case)
ENV JAVA_HOME="/usr/lib/jvm/java-21-openjdk"
ENV PATH="$JAVA_HOME/bin:$PATH"

# Install Terraform
RUN apk add --no-cache terraform

#Install Docker
RUN apk add --no-cache docker

# Install buildx

RUN apk add docker-cli-buildx

# Install Python and Pip
RUN apk add --no-cache python3 py3-pip

# Install AWS CLI
RUN apk add aws-cli-v2

# Add AWS credentials (ensure the credentials file exists)
# ADD ./aws.credentials /root/.aws/credentials

# Switch back to GoCD user


FROM gocd/gocd-agent-wolfi:v24.3.0

USER root

# Install Maven
RUN apk add --no-cache maven

# Install OpenJDK 8 and Java 21
RUN apk add --no-cache openjdk-8 openjdk-21

# Set JAVA_HOME to the default Java version (Java 21 in this case)
ENV JAVA_HOME="/usr/lib/jvm/java-21-openjdk"
ENV PATH="$JAVA_HOME/bin:$PATH"

# Install Terraform
RUN apk add --no-cache terraform

#Install Docker
RUN apk add --no-cache docker

# Install buildx

RUN apk add docker-cli-buildx

# Install Python and Pip
RUN apk add --no-cache python3 py3-pip

# Install AWS CLI
RUN apk add aws-cli-v2

# Add AWS credentials (ensure the credentials file exists)
# ADD ./aws.credentials /root/.aws/credentials

# Switch back to GoCD user

FROM gocd/gocd-agent-wolfi:v24.3.0

USER root

# Install Maven
RUN apk add --no-cache maven

# Install OpenJDK 8 and Java 21
RUN apk add --no-cache openjdk-8 openjdk-21

# Set JAVA_HOME to the default Java version (Java 21 in this case)
ENV JAVA_HOME="/usr/lib/jvm/java-21-openjdk"
ENV PATH="$JAVA_HOME/bin:$PATH"

# Install Terraform
RUN apk add --no-cache terraform

#Install Docker
RUN apk add --no-cache docker

# Install buildx

RUN apk add docker-cli-buildx

# Install Python and Pip
RUN apk add --no-cache python3 py3-pip

# Install AWS CLI
RUN apk add aws-cli-v2

# Add AWS credentials (ensure the credentials file exists)
# ADD ./aws.credentials /root/.aws/credentials

# Switch back to GoCD user

FROM gocd/gocd-agent-wolfi:v24.3.0

USER root

# Install Maven
RUN apk add --no-cache maven

# Install OpenJDK 8 and Java 21
RUN apk add --no-cache openjdk-8 openjdk-21

# Set JAVA_HOME to the default Java version (Java 21 in this case)
ENV JAVA_HOME="/usr/lib/jvm/java-21-openjdk"
ENV PATH="$JAVA_HOME/bin:$PATH"

# Install Terraform
RUN apk add --no-cache terraform

#Install Docker
RUN apk add --no-cache docker

# Install buildx

RUN apk add docker-cli-buildx

# Install Python and Pip
RUN apk add --no-cache python3 py3-pip

# Install AWS CLI
RUN apk add aws-cli-v2

# Add AWS credentials (ensure the credentials file exists)
# ADD ./aws.credentials /root/.aws/credentials

# Switch back to GoCD user

How to solve this ?
ERROR: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?

**Reply by chadlwilson** [reply]
> How to solve this ?
> ERROR: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?

When doing what?

**Reply by chadlwilson** [reply]
If you are trying to run Docker from inside Docker/agent container as part of your builds this is a very complex problem unrelated to GoCD with different approaches and different trade offs. You can read about 'docker in docker' or 'docker out of docker' approaches. You need a docker daemon running inside the agent container as well as outside.

For GoCD there is an agent image based on the official docker dind inage at https://hub.docker.com/r/gocd/gocd-agent-docker-dind which is the easiest way to do this, but you'd be switching from Wolfi to Alpine (both use apk, but package names may differ and Alpine has its own challenges with some tools).

If you only need to build images - but not run them - it is easier, and there are other options to build images without docker and without a daemon, but still not easy.

**Comment by Tanmoy037** [comment]
according to you using docker-dind base image -

```
FROM gocd/gocd-agent-docker-dind:v24.4.0

USER root

# Install Maven
RUN apk add --no-cache maven

# Install OpenJDK 8 and Java 21
RUN apk add --no-cache openjdk8 openjdk21

# Set JAVA_HOME to the default Java version (Java 21 in this case)
ENV JAVA_HOME="/usr/lib/jvm/java-21-openjdk"
ENV PATH="$JAVA_HOME/bin:$PATH"

# Install Terraform
ARG TARGETARCH
ENV TERRAFORM_VERSION=1.5.7
RUN curl -fsSL https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_linux_${TARGETARCH}.zip -o terraform.zip \
    && unzip terraform.zip \
    && mv terraform /usr/local/bin/ \
    && rm terraform.zip

# Install Python and Pip
RUN apk add --no-cache python3 py3-pip

# Install AWS CLI v2
RUN apk add --no-cache aws-cli-v2

# Switch back to GoCD user
# USER go
```

but getting this error in pods - 

kubectl logs -f my-gocd-agent-778bc55495-fzxlv -n gocd
$ sudo /run-docker-daemon.sh
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
dockerd startup failed...
mount: permission denied (are you root?)
Could not mount /sys/kernel/security.
AppArmor detection and --privileged mode might break.
mount: permission denied (are you root?)
/docker-entrypoint.sh: cannot sudo /run-docker-daemon.sh

Please share any solution if you have!

**Reply by chadlwilson** [reply]
Running docker inside kubernetes is even more complicated than normal docker, partly because kubernetes doesn't use docker itself anymore - it uses raw containerd or Cri-O etc. Depends on your kubernetes distribution, security settings and controls, so you'll need to do some reading.

Nevertheless at a minimum look at the container privileges and mounts needed at https://hub.docker.com/r/gocd/gocd-agent-docker-dind 

I can't really help you with docker-in-docker on Kubernetes in arbitrary environments a sonly have experience with EKS (which has some magic to keep Docker-in-docker working). This is a generic Kubernetes problem that is outside the scope of GoCD to solve and you'll need to look elsewhere for your specific Kubernetes environment to find the appropriate pod security settings and configuration that's appropriate for your environment and security position.
